src/Avalonia.Base/AvaloniaObject.cs

```diff
@@ -553,19 +559,44 @@ protected virtual void OnPropertyChanged(AvaloniaPropertyChangedEventArgs e)
         protected bool SetAndRaise<T>(AvaloniaProperty<T> property, ref T field, T value)
         {
             VerifyAccess();
-            if (!object.Equals(field, value))
+            if (!directDelayedSetter.IsNotifying(property))
             {
-                var old = field;
-                field = value;
-                RaisePropertyChanged(property, old, value, BindingPriority.LocalValue);
-                return true;
+                var valueChanged = false;
+                if (!object.Equals(field, value))
+                {
+                    SetAndRaiseCore(property, ref field, value);
+
+                    valueChanged = true;
+                }
+
+                while (directDelayedSetter.HasPendingSet(property))
+                {
+                    SetAndRaiseCore(property, ref field, (T)directDelayedSetter.GetFirstPendingSet(property));
+                    valueChanged = true;
+                }
+                return valueChanged;
             }
-            else
+            else if(!object.Equals(field, value))
+            {
+                directDelayedSetter.AddPendingSet(property, value);
+            }
+            return false;
+        }
+
+        private void SetAndRaiseCore<T>(AvaloniaProperty<T> property, ref T field, T value)
+        {
+            var old = field;
+            field = value;
+
+            using (directDelayedSetter.MarkNotifying(property))
             {
-                return false;
+                RaisePropertyChanged(property, old, value, BindingPriority.LocalValue);
             }
         }
 
+        protected void SetAndRaise<T>(AvaloniaProperty<T> property, Action<T, Action<Action>> setterCallback, T value, Predicate<T> pendingSetCondition = null)
```

Comment:
When would one use this overload of `SetAndRaise` over the other one? I notice it's used by `SelectingItemsControl.SelectedIndex` - why only there?

Also `pendingSetCondition` is never passed into this so `o => pendingSetCondition?.Invoke((T)o) ?? true` is allocating a lambda just to return `true` (at least I think it will cause an allocation as it's capturing a local). Why not allow passing null to `SetAndNotify` if no `pendingSetCondition` is needed?

Reply:
You use this overload when you have manual logic for setting and manually call `RaisePropertyChanged`. If you don't need to control the logic around when you call `RaisePropertyChanged` (i.e. 90% of cases), you just call the other overload.